drivers/clock_control/clock_control_wch_rcc.c

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2024 Michael Hope
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT wch_rcc
+
+#include <stdint.h>
+
+#include <zephyr/arch/cpu.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/clock_control.h>
+
+#include <ch32_rcc.h>
+
+#define WCH_RCC_CLOCK_ID_OFFSET(id) (((id) >> 5) & 0xFF)
+#define WCH_RCC_CLOCK_ID_BIT(id)    ((id) & 0x1F)
+
+struct clock_control_wch_rcc_config {
+	RCC_TypeDef *regs;
+};
+
+static int clock_control_wch_rcc_on(const struct device *dev, clock_control_subsys_t sys)
+{
+	const struct clock_control_wch_rcc_config *config = dev->config;
+	RCC_TypeDef *regs = config->regs;
+	uint8_t id = (uintptr_t)sys;
+
+	*(&regs->AHBPCENR + WCH_RCC_CLOCK_ID_OFFSET(id)) |= 1 << WCH_RCC_CLOCK_ID_BIT(id);
```

Comment:
Maybe something like this?

```suggestion
	uint32_t reg = (uint32_t)&regs->AHBPCENR + WCH_RCC_CLOCK_ID_OFFSET(id)));
  uint32_t val = sys_read32(reg);

  reg |= BIT(WCH_RCC_CLOCK_ID_BIT(id));
  sys_write32(val, reg);
```

Reply:
Thank you - This code suggestion really helps! Pushed now.